components/omega/src/ocn/OceanState.h

```diff
@@ -0,0 +1,126 @@
+#ifndef OMEGA_OCEANSTATE_H
+#define OMEGA_OCEANSTATE_H
+//===-- ocn/OceanState.h - ocean state --------------------*- C++ -*-===//
+//
+/// \file
+/// \brief Contains the state variables for an OMEGA sub-domain
+///
+/// The OceanState class contains the non-tracer prognostic variable data for a
+/// sub-domain of the global horizontal mesh.
+//
+//===----------------------------------------------------------------------===//
+
+#include "DataTypes.h"
+#include "Decomp.h"
+#include "Halo.h"
+#include "HorzMesh.h"
+#include "MachEnv.h"
+
+#include <string>
+
+namespace OMEGA {
+
+/// A class for the ocean prognostic variable information
+
+/// The OceanState class provides a container for the layer thickness,
+/// and normal velocity variables. It contains methods which handle
+/// IO and time level updates.
+class OceanState {
+
+ private:
+   void initParallelIO(Decomp *MeshDecomp);
+
+   void finalizeParallelIO();
+
+   void read();
+
+   void defineIOFields();
+
+   I4 CellDecompR8;
+   I4 EdgeDecompR8;
+   Halo *MeshHalo;
+
+   static OceanState *DefaultOceanState;
+
+   static std::map<std::string, OceanState> AllOceanStates;
+
+ public:
+   // Variables
+   // Since these are used frequently, we make them public to reduce the
+   // number of retrievals required.
+
+   std::string StateFileName;
+   int StateFileID;
+
+   // Sizes and global IDs
+   // Note that all sizes are actual counts (1-based) so that loop extents
+   // should always use the 0:NCellsXX-1 form.
+
+   I4 NCellsOwned; ///< Number of cells owned by this task
+   I4 NCellsAll;   ///< Total number of local cells (owned + all halo)
+   I4 NCellsSize;  ///< Array size (incl padding, bndy cell) for cell arrays
+
+   I4 NEdgesOwned; ///< Number of edges owned by this task
+   I4 NEdgesAll;   ///< Total number (owned+halo) of local edges
+   I4 NEdgesSize;  ///< Array length (incl padding, bndy) for edge dim
+
+   static const I4 MaxTimeLevels = 5; ///< Maximum number of time levels
+
+   I4 NTimeLevels; ///< Number of time levels in state variable arrays
+   I4 NVertLevels; ///< Number of vertical levels in state variable arrays
+
+   // Prognostic variables
+
+   Kokkos::Array<Array2DR8, MaxTimeLevels>
```

Comment:
Do we want the state to always be in double precision or  use  `Array2DReal` to optionally allow single precision ?